### PR TITLE
Add FM Radio stream to AmpliPi

### PR DESCRIPTION
**Pull request by kjk2010**
This pull request adds the ability to use an RTL-SDR-compatible USB dongle as an FM Radio source in AmpliPi. It uses the 'rtl_fm' program as part of the rtl-sdr package.
Current limitations: mono only, no RDS metadata.
RDS metadata can be achieved using https://github.com/windytan/redsea, but I had a lot of issues with stability when using redsea and Python's subprocess.
Stereo might be achievable with https://github.com/jorisvr/SoftFM
HDRadio might be an option as well with https://github.com/theori-io/nrsc5

### CODE REVIEW

**Comment by linknum23** [comment]
Just ordered a USB SDR from Amazon, https://www.amazon.com/RTL-SDR-Blog-RTL2832U-Software-Defined/dp/B011HVUEME,  and am are excited to test this. We will check out the RDS integration with redsea to try to get past some of the issues you experienced.

**Comment by kjk2010** [comment]
That's the same device I have, and it works well with the Pi's in my experience. I'd be very excited if we can get RDS/metadata working reliably with this.

I've attached the Python file I created that runs similarly to runvlc.py to monitor and update a currentSong JSON file with the updated metadata from redsea.

The issue I ran into is after running the Python script that starts rtl_fm, redsea, and aplay for a few minutes, the audio would stop for a few seconds, then play for a second or two, then stop again. While testing, I pulled everything out of the attached script that had to do with parsing the RDS data, but the issue still occurred. It would only happen when tuned to a station with lots of RDS data, which makes me think subprocess isn't keeping up with the redsea data in the pipe, and eventually becomes overwhelmed.

I thought about moving the redsea processing code out of Python and into something else, but with how much data redsea produces, I couldn't output the redsea pipe to a file (without killing the Pi's storage), and didn't want to introduce more dependencies. Hopefully it's just something I overlooked.

Thanks!

[fmradio.zip](https://github.com/micro-nova/AmpliPi/files/6919626/fmradio.zip)

**Comment by linknum23** [comment]
I tried the prototype with RDS decoding out today, in your example there was a sleep(1) that was causing redsea's stderr pipe to overflow. Here's an updated file with some notes and fixes:
[fmradio.py.zip](https://github.com/micro-nova/AmpliPi/files/6927333/fmradio.py.zip)

**Comment by kjk2010** [comment]
I've been testing your update and it's been working perfectly. I'll update the branch with the new code.
I'll add the install of redsea to configure.py, but I'm not sure if this is the only place it needs to be or if I'm adding it correctly...
Thanks!

**Comment by linknum23** [comment]
I need to test this again when I am back in the office tomorrow. It should be ready to merge after that.

**Review by linknum23 — CHANGES_REQUESTED** [review]

File: scripts/configure.py [hunk 1]
```diff
@@ -38,6 +38,16 @@
   'internet_radio' : {
     'apt' : [ 'vlc' ]
   },
+  'fmradio' : {
+    'apt' : [ 'rtl-sdr', 'git', 'build-essential', 'autoconf', 'libsndfile1-dev', 'libliquid-dev' ],
+    'script' : [
```
[the remark this answers is not in the document]
Comment (linknum23) [reply scripts/configure.py:43]: Also I looked at the configure code, looks like the script option was removed awhile back. We will need to add it back for this to work.

**Review by linknum23 — CHANGES_REQUESTED** [review]

File: amplipi/streams.py [hunk 2]
```diff
@@ -750,8 +750,115 @@ def info(self) -> models.SourceInfo:
   def status(self):
     return self.state
 
+class FMRadio:
```
[the remark this answers is not in the document]
Comment (micronova-jb) [reply amplipi/streams.py:753]: I'm currently working on integrating these changes :)

File: streams/fmradio.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+          if "prog_type" in rds:
+            if rds["prog_type"] != latest_info["prog_type"]:
+              latest_info["prog_type"] = rds["prog_type"]
+              update = True
+          if "radiotext" in rds:
+            if rds["radiotext"] != latest_info["radiotext"]:
+              latest_info["radiotext"] = rds["radiotext"]
+              update = True
+            print(f'rt; "{rds["radiotext"]}"')
+          elif "partial_radiotext" in rds:
```
[the remark this answers is not in the document]
Comment (kjk2010) [reply streams/fmradio.py]: I have a local station that only broadcasts in partial radiotext, nothing in radiotext, but they don't have anything worthwhile in the partial radiotext, so I can agree.

File: streams/fmradio.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  latest_info = {
+    'station':'',
+    'callsign':'',
+    'prog_type': '',
+    'radiotext': ''
+  }
+
+  update = False
+
+  rtlfm_args = ['rtl_fm', '-M', 'fm', '-f', '{}M'.format(freq), '-s', '171k', '-A', 'std', '-p', '0', '-l', '0', '-E', 'deemp', '-g', '20', '-F', '9']
```
[the remark this answers is not in the document]
Comment (kjk2010) [reply streams/fmradio.py:65]: This does make it easier, will do.

**Review by kjk2010 — COMMENTED** [review]
Thanks for the comments. I've made some changes and also updated the code to work with the current master.

File: scripts/configure.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (linknum23) [reply scripts/configure.py:43]: I just tried this on our AmpliPi and it looks like a udev rule is not getting installed for rtl_fm and it silently fails. I couldn't find the documentation for how to do this. Can you try adding the udev rule install to the script?

File: scripts/configure.py [hunk 5]
```diff
@@ -38,6 +38,16 @@
...
+      'cd /tmp',
+      'git clone https://github.com/windytan/redsea.git',
```
Comment (linknum23) [inline scripts/configure.py]: Maybe reduce the network load of the git clone?
```suggestion
      'git clone --depth=1 https://github.com/windytan/redsea.git',
```

**Review by linknum23 — CHANGES_REQUESTED** [review]
Updated the streams to the latest changes and made the install conditional because it took awhile on the pi.

Can you look into the udev rule stuff? Once that is added I think this will be good to go.

**Comment by kjk2010** [comment]
> Updated the streams to the latest changes and made the install conditional because it took awhile on the pi.
> 
> Can you look into the udev rule stuff? Once that is added I think this will be good to go.

As a test, would you be able to manually add the udev rules file on the AmpliPi, reboot, and see if that fixes the issue? If so, we could bundle and copy the rtl-sdr.rules file from the scripts folder.

Download and place the rules files into udev rules:
sudo wget https://raw.githubusercontent.com/osmocom/rtl-sdr/master/rtl-sdr.rules -P /etc/udev/rules.d/

**Comment by linknum23** [comment]
👍 I'll test that tomorrow.

**Comment by linknum23** [comment]
Worked great. Here's the commands I used.
```bash
sudo wget https://raw.githubusercontent.com/osmocom/rtl-sdr/master/rtl-sdr.rules -P /etc/udev/rules.d/
sudo udevadm control --reload-rules
sudo udevadm trigger
```

**Comment by kjk2010** [comment]
Sweet, thanks for testing! I'll get configure.py updated here in a few.

**Comment by linknum23** [comment]
Sweet, looks great. Once I'm able to verify this works on the base image I'll merge this in!

**Comment by kjk2010** [comment]
> I tried testing these changes on top of master today and didn't have any luck with the install. I think I missed something. My test is in https://github.com/micro-nova/AmpliPi/tree/fmradio. I didn't push my changes just in case I missed something.

I'm going to wipe my RPi and rebuild it and see if I can duplicate the issue.

I'm curious, on your test rig, does `rtl_test` show any attached devices?
Mine shows this with two plugged in:
pi@amplipi:~$ rtl_test
Found 2 device(s):
  0:  Realtek, RTL2838UHIDIR, SN: 00000001
  1:  Realtek, RTL2838UHIDIR, SN: 00000001

Using device 0: Generic RTL2832U OEM
Detached kernel driver
Found Rafael Micro R820T tuner
Supported gain values (29): 0.0 0.9 1.4 2.7 3.7 7.7 8.7 12.5 14.4 15.7 16.6 19.7                               20.7 22.9 25.4 28.0 29.7 32.8 33.8 36.4 37.2 38.6 40.2 42.1 43.4 43.9 44.5 48.0                               49.6
[R82XX] PLL not locked!
Sampling at 2048000 S/s.
...

**Comment by linknum23** [comment]
Hmm. Once I got into the office I tested this several times on boxes we are planning to ship next week and everything worked fine after a small change to the ALSA config.

**Review by linknum23 — APPROVED** [review]

**Review by linknum23 — APPROVED** [review]

**Comment by linknum23** [comment]
After more testing everything was working good. I updated the docs to show that we support FM Radio now.

**Pull request merged**